### ISSUE TITLE
False positive dependent template name

### ISSUE THREAD

**Issue by cloudhan** [llvm/llvm-project#57019]
The following code snippet fails with
```
<source>:31:5: error: use 'template' keyword to treat 'Ptr' as a dependent template name
  o.Ptr<T>();
    ^
    template 
1 error generated.
ASM generation compiler returned: 1
<source>:31:5: error: use 'template' keyword to treat 'Ptr' as a dependent template name
  o.Ptr<T>();
    ^
    template 
1 error generated.
Execution build compiler returned: 1
```

But if you compile with `-DFIX1` or `-DFIX2`, it will workaround the error and compile smoothly. I don't see the `o` object being something dependent on a template parameter. So the `o.Ptr<T>();` should be fine here. So I think this is a false positive (even MSVC can compile it!). I haven't passed the bar examination of C++, so please kindly point me out if I am wrong here.

https://godbolt.org/z/rv3nTzxvP

```cpp
#include <cstdint>
#include <vector>

class Var final {
public:
  template <typename T>
  T* Ptr() { return nullptr; }
};

class Context {
 public:
  Var* Output(const std::vector<int64_t>&) { return nullptr; }
};

template<typename T>
class Op {
public:
  void Compute(Context* ctx);
  int64_t v;
};

template<typename T>
void Op<T>::Compute(Context* ctx) {
#if   defined(FIX1)
  Var& o = *ctx->Output({this->v});
#elif defined(FIX2)
  auto& o = *ctx->Output({int64_t(this->v)});
#else
  auto& o = *ctx->Output({this->v});
#endif
  o.Ptr<T>();
}

int main() {
  Context ctx{};
  Op<float> o{};
  o.Compute(&ctx);
}
```
